File: app/lib/folders/dtos/folder_element_dto.dart
import 'package:app/core/dtos/base_dto.dart';
import 'package:app/shared/dtos/label_dto.dart';
import 'package:app/user/dtos/user_dto.dart';
import 'package:json_annotation/json_annotation.dart';

part 'folder_element_dto.g.dart';

@JsonSerializable()
class FolderElementDto extends BaseDto {
  @JsonKey(required: true)
  late String name;

  @JsonKey(required: true)
  late LabelDto label;

  @JsonKey(required: true, name: 'image_url')
  late String imageUrl;

  @JsonKey(required: true, name: 'date_uploaded')
  late DateTime dateUploaded;

  @JsonKey(required: true, name: 'date_changed')
  late DateTime dateChanged;

  @JsonKey(required: true, name: 'last_user_change')
  late UserDto lastUserChange;

  FolderElementDto(String id, this.name, this.label, this.imageUrl,
      this.dateUploaded, this.dateChanged, this.lastUserChange)
      : super(id);

  factory FolderElementDto.fromJson(Map<String, dynamic> json) =>
      _$FolderElementDtoFromJson(json);

  Map<String, dynamic> toJson() => _$FolderElementDtoToJson(this);
}

File: backend/src/folder/dto/folder_element_dto.py
from datetime import date

from pydantic import BaseModel

from src.database.tables import FolderElement
from src.shared.dto import LabelDto
from src.user.dto.user_dto import UserDto


class FolderElementDto(BaseModel):
    id: str
    name: str
    label: LabelDto
    image_url: str
    date_uploaded: date
    date_changed: date
    last_user_change: UserDto

    @staticmethod
    def from_database(folder_element: FolderElement):
        return FolderElementDto(
            id=folder_element.id, name=folder_element.name,
            label=LabelDto.from_database(folder_element.label_connection.label),
            image_url=folder_element.image_path,
            date_uploaded=folder_element.date_uploaded,
            date_changed=folder_element.date_changed,
            last_user_change=UserDto.from_database(folder_element.last_user_change)
        )
